server/src/main/java/org/opensearch/index/store/remote/utils/cache/LRUCache.java

```diff
@@ -0,0 +1,423 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.store.remote.utils.cache;
+
+import org.opensearch.common.cache.RemovalListener;
+import org.opensearch.common.cache.RemovalNotification;
+import org.opensearch.common.cache.RemovalReason;
+import org.opensearch.common.cache.Weigher;
+import org.opensearch.index.store.remote.utils.cache.stats.CacheStats;
+import org.opensearch.index.store.remote.utils.cache.stats.DefaultStatsCounter;
+import org.opensearch.index.store.remote.utils.cache.stats.StatsCounter;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.BiFunction;
+
+/**
+ * LRU implementation of {@link RefCountedCache}. As long as {@link Node#refCount} greater than 0 then node is not eligible for eviction.
+ * So this is best effort lazy cache to maintain capacity.
+ *
+ * @see RefCountedCache
+ *
+ * @opensearch.internal
+ */
+class LRUCache<K, V> implements RefCountedCache<K, V> {
```

Comment:
Why not use `LRUMemoryCache` from [apache commons jcs](Why not use org.apache.commons.jcs) instead of reinventing here?

Reply:
Yes it make sense to not reinvent the wheel and write our own cache from scratch. Reasons why we went for custom cache implementation as first release is here https://github.com/opensearch-project/OpenSearch/issues/4964#issuecomment-1421689586 . Basically we have evaluated Guava/Caffeine cache and they are not open for extension for implement our custom eviction logic. Apache Commons JCS look promising but I was not able to quickly find path forward to use it. So I've created this follow up task https://github.com/opensearch-project/OpenSearch/issues/6225 to address this concern .

Also for heap usage impact I've created this task https://github.com/opensearch-project/OpenSearch/issues/6227 as fast follow up.